### DISCUSSION TITLE
[FEATURE] Provide better Error messages, with a priority on REPLs

### DISCUSSION THREAD

**Discussion by yannickfunk** [pyscript/pyscript#1175] [Proposals]
**Is your feature request related to a problem? Please describe.**
Currently, when running invalid code in a REPL, the error message from pyodide is directly propagated. Something like:
```
JsException(PythonError: Traceback (most recent call last): File "/lib/python3.10/site-packages/_pyodide/_base.py", line 429, in eval_code .run(globals, locals) File "/lib/python3.10/site-packages/_pyodide/_base.py", line 300, in run coroutine = eval(self.code, globals, locals) File "", line 2, in NameError: name 'adwwda' is not defined )
```
shows up.

**Describe the solution you'd like**
A desirable solution would be intuitive Error messages which hide the internals of pyodide, like in Jupyter Notebooks e.g.
```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
Input In [1], in <cell line: 1>()
----> 1 adwwda

NameError: name 'adwwda' is not defined
```

**Additional context**
This would be a more complex feature to implement since it involves the base component, all components which implement `eval()` and the python side with `OutputCtxManager` and `OutputManager`. If there are already planned efforts to do this, I would be happy to help, otherwise I would also be happy to start investigating on possibilities here.

**Comment by yannickfunk** [comment]
To achieve this we would need to think of how to accomplish the following:

- [ ] Distinguish user written python code from pyscript internal python code (e.g. calls to output manager)
- [ ] Ability to map dependencies between repl cells. (Somehow track which part of the executed python code originates from which repl cell)
- [ ] Separate pyodide stacktrace from user code stacktrace
- [ ] Implement proper rendering for the python errors

Any more suggestions for what we would need? @philippjfr @marimeireles @fpliger

**Comment by marimeireles** [comment]
Hey @yannickfunk, thanks for raising the issue! 

> Separate pyodide stacktrace from user code stacktrace

My 2 cents on this... is that I was checking on this yesterday and I think the error output is coming directly from Pyodide. So doing something like what you're proposing might not be possible? I'm not sure, maybe you looked into it deeper than I did. Fabio certainly has a precise input on this :)

But one solution I thought, that might not be the best, but might be pretty easy to implement is to just parse this string and show a pretty output to the user.

(Fabio is out for the next few days but must be back soon!)

**Comment by yannickfunk** [comment]
@marimeireles  Yeah the output comes directly from pyodide, my intention here was to parse this output and hide the pyodide internals from the user. That's what I meant with 

> Separate pyodide stacktrace from user code stacktrace


E.g for this repl:
![image](https://user-images.githubusercontent.com/44268911/171702620-387705f4-75b1-4fdc-ae5a-4f7011928ae8.png)

The error would be
```
Traceback (most recent call last):
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 429, in eval_code
    .run(globals, locals)
  File "/lib/python3.10/site-packages/_pyodide/_base.py", line 300, in run
    coroutine = eval(self.code, globals, locals)
  File "<exec>", line 2, in <module>
  File "<exec>", line 3, in a
ZeroDivisionError: division by zero
```

Where the user ideally should only see the last 3 lines (also containing the right lines and cell information)

**Comment by marimeireles** [comment]
Ahh ok. Sorry I misunderstood you.
Yeah, I think it's a good approach.
I feel like Fabio would like this idea too, but maybe he knows something about the internals that would offer a better solution.

I guess this might be the relevant object we wanna parse: https://github.com/pyscript/pyscript/blob/fa28b218ecea602236e536363771a2f4fa1e8ae5/pyscriptjs/src/components/base.ts#L184

**Comment by fpliger** [comment]
Sorry for being late to the party. The feature request here is definitely 👍  on my side :)
